### PR TITLE
#167164992 follow and unfollow user feature

### PR DESCRIPTION
**Pull request by OvieMudi**
#### What does this PR do?
- authenticated users can follow each other
- authenticated users can unfollow each other
#### Description of Task to be completed?
- Ensure the endpoint `/api/v1/profile/<userId>/follow` create a new user in the database when provided with appropriate fields

#### How should this be manually tested?
To test manually on your local computer:
-  run the following commands.

```
# Clone this repository
$ git clone https://github.com/andela/dahlia-ah-backend.git

# Navigate to the application folder
$ cd dahlia-ah-backend

# Checkout the branch
$ git checkout feature/167164992/users-follow-each-other

# Install dependencies
$ npm install

# Start the app in development mode
$ npm run dev:start
```
- on Postman make an authorized request to `/api/v1/profile/<userId>/follow`

Note: See the screenshot below on how the  request/response should look
#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#167164992](https://www.pivotaltracker.com/story/show/167164992)

#### Screenshots
Screenshot 2019-08-08 at 12 08 45 PM
Screenshot 2019-08-08 at 12 09 32 PM
Screenshot 2019-08-08 at 12 13 11 PM

#### API Flowchart
Follow and unfollow features

### CODE REVIEW

**Review by Nerocodes — APPROVED** [review]
LGTM

**Comment by Ukhu** [comment]
Kindly add a screenshot of your flowchart for this task and increase test coverage.

**Comment by OvieMudi** [comment]
> Kindly add a screenshot of your flowchart for this task and increase test coverage.

Thank you very much for this observation. I'm on it.

**Review by allebd — COMMENTED** [review]
I observed that the flowchart attached does not describe the follow user feature

**Comment by OvieMudi** [comment]
> I observed that the flowchart attached does not describe the follow user feature

Thank you very much for that observation. That is obviously a mixup. I'll make changes as soon as I can.

**Review by benisi — APPROVED** [review]
LGTM

**Review by profmcdan — CHANGES_REQUESTED** [review]
Implement the un-follow feature

File: src/controllers/userController.js [hunk 1]
```diff
@@ -139,6 +139,81 @@ const editProfile = async (request, response) => {
...
+    }
+
+    await Follower.create({
+      followerId: authUserId,
+      followeeId: userId
+    });
+
+    const user = await findUser(userId);
+    const { bio, avatarUrl } = user;
+    const data = {
```
Comment (Ukhu) [inline src/controllers/userController.js:187]: you can return only the userId and the following status instead

**Review by profmcdan — CHANGES_REQUESTED** [review]
Attend to your reviews

File: src/controllers/userController.js [hunk 2]
```diff
@@ -158,6 +158,78 @@ const listUsers = async (request, response) => {
...
+   */
+const follow = async (req, res) => {
+  const { userId } = req.params;
+  const authUserId = req.user.id;
+  if (userId === authUserId) {
+    return responseMessage(res, 403, { error: 'You cannot follow yourself' });
+  }
+
+  try {
+    const isFollower = await findFollower(userId, authUserId);
```
Comment (joel-ace) [inline src/controllers/userController.js]: don't you think follower is a better variable name? `isFollower` kinda makes it seem like I'm expecting a boolean response

File: src/controllers/userController.js [hunk 3]
```diff
@@ -158,6 +158,78 @@ const listUsers = async (request, response) => {
...
+   * @return {Object} - custom response
+   */
+const unfollow = async (req, res) => {
+  const { userId } = req.params;
+  const authUserId = req.user.id;
+  if (userId === authUserId) {
+    return responseMessage(res, 403, { error: 'operation not allowed' });
+  }
+  try {
+    const isFollower = await findFollower(userId, authUserId);
```
Comment (joel-ace) [inline src/controllers/userController.js]: same comment about isFollower above

File: src/controllers/userController.js (hunk 2 above)
Comment (OvieMudi) [reply src/controllers/userController.js]: Yeah, I agree.

File: src/controllers/userController.js (hunk 3 above)
Comment (OvieMudi) [reply src/controllers/userController.js]: I just made the changes

**Pull request merged**